### PR TITLE
single Makefile.inc and build.sh for scotch and ptscotch

### PR DESCRIPTION
**Pull request by minrk**
only include deviations from default in platform Makefiles

build number is bumped because -DSCOTCH_PTHREAD is restored, which had been removed.

### CODE REVIEW

**Comment by dalcinl** [comment]
Sorry about the removal of `-DSCOTCH_PTHREAD `.  This bug slipped from my ongoing work on ptscotch. For the parallel variant, IMHO we should not use that flag. Too bad that scotch only allows compile-time configuration of threads.

**Comment by dalcinl** [comment]
@minrk Either on Linux or macOS we use GNU Make. Then, why don't we just provide `Makefile.inc` and inside it do:

```make
ifeq ($(shell uname ),Linux)
    ABC = xyz
else
    ABC = xyz
endif
```

File: recipe/Makefile.inc.Darwin [hunk 1]
```diff
@@ -1,22 +1,8 @@
...
-LDFLAGS         = -L${PREFIX}/lib -lz -lm -pthread -Wl,-headerpad_max_install_names
-CP              = cp
-LEX             = flex -Pscotchyy -olex.yy.c
-LN              = ln
-MKDIR           = mkdir
-MV              = mv
-RANLIB          = ranlib
-YACC            = bison -pscotchyy -y -b y
+CLIBFLAGS      := $(CLIBFLAGS) -fPIC
+LDFLAGS        := $(LDFLAGS) -Wl,-headerpad_max_install_names
```
Comment (dalcinl) [inline recipe/Makefile.inc.Darwin]: Maybe we are doing something wrong here? Isn't `conda build` supposed to pass `-Wl,-headerpad_max_install_names` in the `LDFLAGS` environ var?

File: recipe/Makefile.inc.common [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+MAKE            = make
+AR              = ar
+ARFLAGS         = -ruv
+CAT             = cat
+CCS             = gcc
+CCP             = mpicc
+CCD             = gcc
+CFLAGS          = -I${PREFIX}/include -O3 -DIDXSIZE64 -Drestrict=__restrict -DCOMMON_FILE_COMPRESS_GZ -DCOMMON_RANDOM_FIXED_SEED -DSCOTCH_RENAME -DCOMMON_PTHREAD -DCOMMON_PTHREAD_BARRIER -DSCOTCH_PTHREAD
+CLIBFLAGS       =
+LDFLAGS         = -L${PREFIX}/lib -lz -lm -pthread
```
Comment (dalcinl) [inline recipe/Makefile.inc.common]: Maybe we should do `LDFLAGS := ${LDFLAGS} -L${PREFIX}/lib -lz -lm -pthread`?

File: recipe/build.sh [hunk 3]
```diff
@@ -1,9 +1,10 @@
 #!/bin/sh
 
+cp $RECIPE_DIR/Makefile.inc.common src/Makefile.inc.common
 cp $RECIPE_DIR/Makefile.inc.$(uname) src/Makefile.inc
 
 cd src/
-make esmumps | tee make.log 2>&1
+make -j ${NUM_CPUS} esmumps | tee make.log 2>&1
```
Comment (dalcinl) [inline recipe/build.sh]: Are you sure all makefiles are properly written to support parallel builds?

Comment (minrk) [reply recipe/build.sh]: I've just tested and they certainly aren't! Removed.

File: recipe/Makefile.inc [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+EXE             =
+LIB             = .a
+OBJ             = .o
+
+MAKE            = make
+AR              = ar
+ARFLAGS         = -ruv
+CAT             = cat
+CCS            := ${CC}
```
Comment (dalcinl) [inline recipe/Makefile.inc]: I think you hae to do it this way:

```
CCS = ${CC}
CCP = ${CCS}
CCD = ${CCP}
```

Then, in `build.sh`, and only for `ptscotch`, `export CCP=mpicc`.

Comment (dalcinl) [reply recipe/Makefile.inc]: And I think that we no longer need `:=` assignements.

**Comment by dalcinl** [comment]
@minrk I think you have parts of #16. The install step for `ptscotch` has to be different, otherwise you install both the parallel and sequential headers/libraries. That does not look right.

File: recipe/build.sh [hunk 5]
```diff
@@ -1,9 +1,29 @@
...
-cp $RECIPE_DIR/Makefile.inc.$(uname) src/Makefile.inc
+env | sort
+cp $RECIPE_DIR/Makefile.inc src/Makefile.inc
+
+export CFLAGS="${CFLAGS} -I${PREFIX}/include -O3 -DIDXSIZE64 -Drestrict=__restrict -DCOMMON_FILE_COMPRESS_GZ -DCOMMON_RANDOM_FIXED_SEED -DSCOTCH_RENAME -DCOMMON_PTHREAD -DCOMMON_PTHREAD_BARRIER"
+export LDFLAGS="${LDFLAGS} -lz -lm -pthread"
+
+if [[ "$(uname)" == "Darwin" ]]; then
+    export CFLAGS="${CFLAGS} -DCOMMON_PTHREAD_BARRIER -DCOMMON_TIMING_OLD"
+    export CLIBFLAGS="${CLIBFLAGS} -fPIC"
```
Comment (dalcinl) [inline recipe/build.sh]: We need `-fPIC` both in Linux and macOS. Otherwise we may not be able to build a sharedlib PETSc.

**Comment by dalcinl** [comment]
@minrk Looking at your comment in #16, you are definitely right, ptscotch does not actually depend on scotch. The problem is that when you build ptscotch, you also get the scotch libraries and headers in the install tree. So I expect that some third-party libraries (that's the case of PETSc) would expect both the scotch and ptscotch variants at configure time. Making ptscotch depend on scotch would make things simpler. But of course we can just update the PETSc recipe to depend on both scotch and ptscotch.

**Comment by dalcinl** [comment]
We also have a problem with the `esmumps.h` header file, there is no `ptesmumps.h` header, but we have `libesmumps.a` vs `libptesmumps.a`. Which package should own `esmumps.h` ?

File: recipe/build.sh (hunk 5 above)
Comment (dalcinl) [reply recipe/build.sh]: Maybe just seeting `CLIBFLAGS = -fPIC` in `Makefile.inc` is good enough?

**Comment by minrk** [comment]
Should the ptscotch install exclude the scotch libraries? I would think that installing ptscotch and scotch would be mutually exclusive, especially since the scotch build flags should be different (e.g. `-DSCOTCH_PTHREAD` for serial, but not for parallel)

**Comment by dalcinl** [comment]
Yes, for sure the ptscotch install exclude the scotch libraries. But our `build.sh` script was not doing that, we don't `make install/ptinstall`, just `cp ... $PREFIX/<include,bin,lib> `!. My changes in #16 fix that issue.

**Comment by dalcinl** [comment]
Just to make things clear: Right now, the ptscotch package in anaconda.org contains sequential and parallel libraries, with the exceptions of `libesmumps.a`. That was the cause of the previous failure in conda-forge/petsc-feedstock#18 (now it is passing, after a quick test adding both scotch and ptscotch as dependencies).

However, installing both scotch and ptscotch together in a conda environment should work, all libraries have different names, the same for headers (with the exception of `esmumps.h`).

**Comment by dalcinl** [comment]
@minrk Oh! Hold on, `ptscotch.h` includes `scotch.h` !!

**Pull request closed without merging**